sim-lib/src/lib.rs

```diff
@@ -826,41 +888,53 @@ async fn produce_events<N: DestinationGenerator + ?Sized, A: PaymentGenerator +
                     },
                     Err(e) => {
                         log::error!("Could not get amount for {source} -> {destination}: {e}. Please report a bug!");
-                        break;
+                        return Err(SimulationError::PaymentGenerationError(e));
                     },
                 };
 
                 log::debug!("Generated random payment: {source} -> {}: {amount} msat.", destination);
 
                 // Send the payment, exiting if we can no longer send to the consumer.
                 let event = SimulationEvent::SendPayment(destination.clone(), amount);
-                if let Err(e) = sender.send(event).await {
+                if let Err(e) = sender.send(event.clone()).await {
                     log::debug!(
                         "Stopped random producer for {amount}: {source} -> {destination}. Consumer error: {e}.",
                     );
-                    break;
+                    return Err(SimulationError::MpscChannelError { msg: format!("Error sending simulation event {event:?} to consumer"), error_str: e.to_string() });
                 }
+
             },
         }
     }
 
     log::debug!("Stopped random activity producer {source}.");
-    shutdown.trigger();
+    
+    Ok(())
```

Comment:
Likewise here replace break above with a return?

Works well with the option of moving start/stop logging to the task spawn site.